### ISSUE TITLE
Phenomenally slow write performace (< 2MB/sec) on MooseFS

### ISSUE THREAD

**Issue by onlyjob** [rfjakob/gocryptfs#742]
Gocryptfs v2.3.1 (mounted with `-noprealloc`) demonstrates phenomenally slow write performance on network file system {^1}: less than 2 MB per second, which is far worse than `encfs`. CPU is not congested and `gocryptfs` process uses less than 20% CPU.

Linux 6.0.12 x86_64 Debian GNU/Linux; i7-4820K CPU @ 3.70GHz

{^1}: MooseFS -- perhaps the fastest distributed network file system. In my case, running on SSD-backed storage, on 10 Gbps network.

**Comment by rfjakob** [comment]
Hmm, how many MB/s does EncFS get?

**Comment by onlyjob** [comment]
I've spent most of the day benchmarking and trying to make a useful reproducer. To my surprise, in every test GocryptFS outperformed EncFS and in the end I've realised that the issue was only reproducible on large
directory structure with 230,000+ small-ish files (Average size: 94091.43)
where I was getting consistent ~1MB/sec write speed on GocryptFS in `rsync`.

Which is, of course, to be expected and the remedy was to run few `rsync` processes in parallel.
(Many small files are stressful for all file systems.)

So in the end Gocryptfs performed very well. It looks that EncFS might be marginally faster with writing small files but I'll have to do more testing to confirm that for sure.

My apologies for the noise and thanks for your patience.

Below are my test results in which both EncFS and GocryptFS are located on MooseFS-backed network share.
TLDR: GocryptFS wins. ;)

---

#### 1GiB/urandom to EncFS: 

```
$ ddrescue -D --size 1GiB  /dev/urandom ./testfile.2kill
GNU ddrescue 1.23
Press Ctrl-C to interrupt
     ipos:    1073 MB, non-trimmed:        0 B,  current rate:   7667 kB/s
     opos:    1073 MB, non-scraped:        0 B,  average rate:  26188 kB/s
non-tried:        0 B,  bad-sector:        0 B,    error rate:       0 B/s
  rescued:    1073 MB,   bad areas:        0,        run time:         40s
pct rescued:  100.00%, read errors:        0,  remaining time:         n/a
                              time since last successful read:         n/a
Finished
```

#### 1GiB/urandom to GocryptFS: 

```
$ ddrescue -D --size 1GiB  /dev/urandom ./testfile.2kill
GNU ddrescue 1.23
Press Ctrl-C to interrupt
     ipos:    1073 MB, non-trimmed:        0 B,  current rate:    196 kB/s
     opos:    1073 MB, non-scraped:        0 B,  average rate:  71582 kB/s
non-tried:        0 B,  bad-sector:        0 B,    error rate:       0 B/s
  rescued:    1073 MB,   bad areas:        0,        run time:         14s
pct rescued:  100.00%, read errors:        0,  remaining time:         n/a
                              time since last successful read:         n/a
Finished
```

### Let's experiment with data directory on local SSD:

```
$ du -sh /mnt/01/tst/00
41G     /mnt/01/tst/00

## Show number of files and their average size:
$ find /mnt/01/tst/00 -type f -printf '%s\n' | awk '{s+=$0} END {printf "Count: %u\nAverage size: %.2f\n", NR, s/NR}'
Count: 1847
Average size: 23660260.81
```

`/tmp/qqq` is EncFS; `/tmp/qqqq` is GocryptFS.

#### Upload local data directory to EncFS:

```
$ time rsync -rtl --info progress2,stats --no-i-r /mnt/01/tst/00/ /tmp/qqq/2kill.test/  ## data to EncFS
 43,700,501,707 100%   16.83MB/s    0:41:16 (xfr#1847, to-chk=0/2028)

sent 43,711,316,279 bytes  received 35,692 bytes  17,643,330.77 bytes/sec
total size is 43,700,501,707  speedup is 1.00

real    41m17.095s
user    0m22.405s
sys     1m53.442s
```

#### Upload local data directory to GocryptFS:

```
$ time rsync -rtl --info progress2,stats --no-i-r /mnt/01/tst/00/ /tmp/qqqq/2kill.test/  ## data to GocryptFS
 43,700,501,707 100%   56.54MB/s    0:12:17 (xfr#1847, to-chk=0/2028)

sent 43,711,316,279 bytes  received 35,692 bytes  59,269,629.79 bytes/sec
total size is 43,700,501,707  speedup is 1.00

real    12m18.180s
user    0m19.443s
sys     1m8.514s
```

#### Copy data from EncFS to local SSD

```
$ time rsync -rtl --info progress2,stats --no-i-r /tmp/qqq/2kill.test/ /mnt/01/tst/00.from_EncFS/ ## EncFS to Local SSD
 43,700,501,707 100%   50.17MB/s    0:13:50 (xfr#1847, to-chk=0/2028)

sent 43,711,316,200 bytes  received 35,692 bytes  52,569,274.67 bytes/sec
total size is 43,700,501,707  speedup is 1.00

real    13m51.389s
user    0m19.260s
sys     1m6.489s
```

#### Copy data from GocryptFS to local SSD

```
$ time rsync -rtl --info progress2,stats --no-i-r /tmp/qqqq/2kill.test/ /mnt/01/tst/00.from_GocryptFS/ ## GocryptFS to Local SSD 43,700,501,707 100%   64.05MB/s    0:10:50 (xfr#1847, to-chk=0/2028)

sent 43,711,316,200 bytes  received 35,692 bytes  66,888,067.16 bytes/sec
total size is 43,700,501,707  speedup is 1.00

real    10m53.288s
user    0m18.509s
sys     1m3.089s
```


### Let's test some network-to-network data copy

#### Copy data dir from GocryptFS to GocryptFS

```
$ time rsync -rtl --info progress2,stats --no-i-r /tmp/qqqq/2kill.test/ /tmp/qqqq/2kill.test_from_GocryptFS/ ## GocryptFS to GocryptFS
 43,700,501,707 100%   34.95MB/s    0:19:52 (xfr#1847, to-chk=0/2028)

sent 43,711,316,200 bytes  received 35,692 bytes  36,593,848.38 bytes/sec
total size is 43,700,501,707  speedup is 1.00

real    19m54.049s
user    0m21.089s
sys     1m10.332s
```

#### Copy data dir from EncFS to EncFS

```
$ time rsync -rtl --info progress2,stats --no-i-r /tmp/qqq/2kill.test/ /tmp/qqq/2kill.test_from_EncFS/ ## EncFS to EncFS 43,700,501,707 100%   13.30MB/s    0:52:13 (xfr#1847, to-chk=0/2028)

sent 43,711,316,200 bytes  received 35,692 bytes  13,945,239.08 bytes/sec
total size is 43,700,501,707  speedup is 1.00

real    52m14.051s
user    0m22.947s
sys     2m0.603s
```

#### Copy data dir from GocryptFS to EncFS

```
$ time rsync -rtl --info progress2,stats --no-i-r /tmp/qqqq/2kill.test/ /tmp/qqq/2kill.test_from_GocryptFS/ ## GocryptFS to EncFS
 43,700,501,707 100%   14.47MB/s    0:48:00 (xfr#1847, to-chk=0/2028)

sent 43,711,316,200 bytes  received 35,692 bytes  15,153,874.81 bytes/sec
total size is 43,700,501,707  speedup is 1.00

real    48m4.843s
user    0m23.261s
sys     2m3.917s
```

#### Copy data dir from EncFS to GocryptFS

```
$ time rsync -rtl --info progress2,stats --no-i-r /tmp/qqq/2kill.test/ /tmp/qqqq/2kill.test_from_EncFS/ ## EncFS to GocryptFS
 43,700,501,707 100%   30.69MB/s    0:22:37 (xfr#1847, to-chk=0/2028)

sent 43,711,316,200 bytes  received 35,692 bytes  32,152,520.70 bytes/sec
total size is 43,700,501,707  speedup is 1.00

real    22m39.346s
user    0m21.278s
sys     1m12.510s
```
